### PR TITLE
22 angular spectra

### LINKED ISSUE #22: Angular spectra
Implement the calculation of the angular spectra in the Limber approximation.

### PR DESCRIPTION
**Pull request by marcobonici**
This PR introduce the computation of the twopt in the Limber approximation.

### CODE REVIEW

File: cloelite/cosmology/jax_cosmology.py [hunk 1]
```diff
@@ -576,6 +577,14 @@ def nonlinear_matter_power_spectrum(self, ks, zs):
         """
         return self.halofit(ks, zs)
 
+    def nonlinear_matter_power_spectrum_limber_grid(self, z_l, ks, zs, ells):
```
Comment (gcanasherrera) [inline cloelite/cosmology/jax_cosmology.py:578]: Include docs

File: cloelite/cosmology/jax_cosmology.py [hunk 2]
```diff
@@ -772,3 +781,10 @@ def _difftrapn(function, interval, numtraps):
     points = lox + h * np.arange(0, numtosum)
     s = np.sum(function(points))
     return s
+
+@jax.jit
+def Pkl_interp(k_l, z_l, ks, zs, Pk):
```
Comment (gcanasherrera) [inline cloelite/cosmology/jax_cosmology.py:784]: include docs

File: cloelite/observables/two_point.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            TypeError("The types of the perturbations of the two tracers is not compatible!")
+
+        self.tracer1 = tracer1
+        self.tracer2 = tracer2
+
+class AngularTwoPoint(TwoPoint):
+    def __init__(self, tracer1 : Tracer, tracer2 : Tracer):
+        super().__init__(tracer1, tracer2)
+
+    def get_Cl(self):
```
Comment (gcanasherrera) [inline cloelite/observables/two_point.py:30]: could you promote ell and ks to the args?

File: cloelite/observables/two_point.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        #still have to include weights, basically we are doing unnormalized trapz
+        return result
+
+@jax.jit
+def Cl_integration(WT1, WT2, Pkl, H, chi2):
+    return np.einsum('iz,jz,lz,z,z->lij', WT1, WT2, Pkl, 1/H, 1/chi2)
+
+@jax.jit
+def Pkl_interp(k_l, z_l, ks, zs, Pk):
+    return 10**interpax.interp2d(jax.numpy.log10(k_l), z_l, jax.numpy.log10(ks), zs,
```
Comment (gcanasherrera) [inline cloelite/observables/two_point.py:56]: not sure why you import jax.numpy as np and then you do not use it, but all good.

File: cloelite/observables/two_point.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        result = Cl_integration(WT1, WT2, Pkl, H, chi2)
+        #still have to include weights, basically we are doing unnormalized trapz
+        return result
+
+@jax.jit
+def Cl_integration(WT1, WT2, Pkl, H, chi2):
+    return np.einsum('iz,jz,lz,z,z->lij', WT1, WT2, Pkl, 1/H, 1/chi2)
+
+@jax.jit
+def Pkl_interp(k_l, z_l, ks, zs, Pk):
```
Comment (gcanasherrera) [inline cloelite/observables/two_point.py:55]: I would make this a private method

File: cloelite/observables/two_point.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        Pkl = self.tracer1.perturbations.nonlinear_matter_power_spectrum_limber_grid(zs_calc, ks, zs_calc, ells)
+
+        WT1 = self.tracer1.get_window(zs_calc)
+        WT2 = self.tracer2.get_window(zs_calc)
+        result = Cl_integration(WT1, WT2, Pkl, H, chi2)
+        #still have to include weights, basically we are doing unnormalized trapz
+        return result
+
+@jax.jit
+def Cl_integration(WT1, WT2, Pkl, H, chi2):
```
Comment (gcanasherrera) [inline cloelite/observables/two_point.py:51]: I would promote this to a private method

**Review by gcanasherrera — COMMENTED** [review]
Only minor things: can be addressed in follow up tasks

File: cloelite/cosmology/jax_cosmology.py (hunk 1 above)
Comment (marcobonici) [reply cloelite/cosmology/jax_cosmology.py:578]: Follow-up task.

File: cloelite/cosmology/jax_cosmology.py (hunk 2 above)
Comment (marcobonici) [reply cloelite/cosmology/jax_cosmology.py:784]: Follow-up task.

File: cloelite/observables/two_point.py (hunk 4 above)
Comment (marcobonici) [reply cloelite/observables/two_point.py:56]: Because I am stupid :)

**Pull request merged**